### DISCUSSION TITLE
Interfacing ePaper HAT v2 and Arduino Mega2560

### DISCUSSION THREAD

**Discussion by nlk-limoss** [Bodmer/TFT_eSPI#2160] [Compatible displays and setup files]
I am working on a project using ePaper 2.13inch HAT v2 display and Arduino Mega2560. I would like to use the TFT_eSPI library in tandem with the LVGL libraries. Listed below are 2 issues I would like clarity on.

1. What drivers would be recommended for the Mega2560?  
2. The V2 introduces changes in syntax for example you have epd.Clear() encapsulating both epd.ClearFrameMemory(0xFF); 

epd.DisplayFrame().
**Snippet of epd2in13.h**
```
class Epd : EpdIf {
public:
    unsigned long width;
    unsigned long height;

    Epd();
    ~Epd();
    int  Init(const unsigned char* lut);
    void SendCommand(unsigned char command);
    void SendData(unsigned char data);
    void WaitUntilIdle(void);
    void Reset(void);
    void SetFrameMemory(
        const unsigned char* image_buffer,
        int x,
        int y,
        int image_width,
        int image_height
    );
    void SetFrameMemory(const unsigned char* image_buffer);
    void ClearFrameMemory(unsigned char color);
    void DisplayFrame(void);
    void Sleep(void);

private:
    unsigned int reset_pin;
    unsigned int dc_pin;
    unsigned int cs_pin;
    unsigned int busy_pin;
    const unsigned char* lut;

    void SetLut(const unsigned char* lut);
    void SetMemoryArea(int x_start, int y_start, int x_end, int y_end);
    void SetMemoryPointer(int x, int y);
};

#endif /* EPD2IN13_H */
```

**Snippet of epd2in13v2.h**
```
class Epd : EpdIf {
public:
    unsigned long width;
    unsigned long height;

    Epd();
    ~Epd();
    int  Init(char Mode);
    void SendCommand(unsigned char command);
    void SendData(unsigned char data);
    void WaitUntilIdle(void);
    void Reset(void);
    void Clear(void);
    void Display(const unsigned char* frame_buffer);
    void DisplayPartBaseImage(const unsigned char* frame_buffer);
    void DisplayPart(const unsigned char* frame_buffer);
    void ClearPart(void);
    
    void Sleep(void);
private:
    unsigned int reset_pin;
    unsigned int dc_pin;
    unsigned int cs_pin;
    unsigned int busy_pin;
};

#endif /* EPD1IN54B_H */
```

Is there any support for this v2 implementation?

**Comment by Bodmer** [comment]
This library is not suited to the Mega, it is really written for more powerful 32 bit processors with more RAM and FLASH.

I think this library is more suited to you needs:
https://github.com/ZinggJM/GxEPD

It has more up-to-date drivers for the current generation of ePaper displays.

I have not tried to use LVGL with an ePaper display, I found it a very memory hungry library so the Mega mas struggle with anything more than a simple sketch.

The ePaper support in this library is rather old and remains to support users legacy sketches. I will probably deprecate the use with ePaper as it will take too much effort to maintain compatibility with the increasing range of ePaper displays.

**Reply by nlk-limoss** [reply]
Hi @Bodmer,
Thanks for that piece of information. 

I went ahead and tried with [GxEPD2](https://github.com/ZinggJM/GxEPD2) (recommended on the https://github.com/ZinggJM/GxEPD as a better alternative for newer users and updated drivers). I managed to at least get some text and images displayed. Thanks for suggesting.

Yes, LVGL does tend to be a memory hungry library. Nethertheless, my goal is to at least define a functional toolchain or methodology that can include LVGL as it is required for a project I am working on. They recommended your library so that is why I was so keen on making it work.

At least there is a library that works, now I turn my attention to working with LVGL again.
Thanks again for the suggestion.

Regards
